src/backend/distributed/planner/local_plan_cache.c

```diff
@@ -99,6 +96,66 @@ CacheLocalPlanForShardQuery(Task *task, DistributedPlan *originalDistributedPlan
 }
 
 
+/*
+ * GetLocalShardQuery is a helper function which generates the local shard
+ * query based on the jobQuery.
+ *
+ * It is not guaranteed to have consistent attribute numbers on the shards
+ * and on the shell (e.g., distributed/reference tables) due to DROP COLUMN
+ * commands. To avoid any edge cases due to such discrepancies, we first
+ * deparse the jobQuery with the tables replaced to shards, and parse the
+ * query string back. This is normally a very expensive operation, however
+ * we only do it once per cached local plan, which is acceptable.
+ */
+static Query *
+GetLocalShardQuery(Query *jobQuery, Task *task, ParamListInfo orig_paramListInfo)
+{
+	StringInfo localShardSqlQuery = makeStringInfo();
+	if (jobQuery->commandType == CMD_INSERT)
+	{
+		/*
+		 * For INSERT queries we cannot use pg_get_query_def. Mainly because we
+		 * cannot run UpdateRelationToShardNames on an INSERT query. This is
+		 * because the PG deparsing logic fails when trying to insert into a
+		 * RTE_FUNCTION (which is what will happen if you call
+		 * UpdateRelationToShardNames).
+		 */
+		deparse_shard_query(jobQuery, task->anchorDistributedTableId, task->anchorShardId,
+							localShardSqlQuery);
+	}
+	else
+	{
+		UpdateRelationToShardNames((Node *) jobQuery, task->relationShardList);
+		pg_get_query_def(jobQuery, localShardSqlQuery);
+	}
+
+	ereport(DEBUG5, (errmsg("Local shard query that is going to be cached: %s",
+							localShardSqlQuery->data)));
+
+	/* for parsing back, we need to know the parameters */
+	Oid *parameterTypes = NULL;
+	int numParams = 0;
+	if (orig_paramListInfo != NULL)
+	{
+		/* not used anywhere, so declare here */
+		const char **parameterValues = NULL;
+		ParamListInfo paramListInfo = copyParamList(orig_paramListInfo);
+		ExtractParametersForLocalExecution(paramListInfo, &parameterTypes,
+										   &parameterValues);
+		numParams = paramListInfo->numParams;
+	}
+
+	Query *localShardQuery =
+		ParseQueryString(localShardSqlQuery->data, parameterTypes, numParams);
+
+	/* we don't need the sql statements anymore */
+	pfree(localShardSqlQuery->data);
+	pfree(localShardSqlQuery);
+
+	return localShardQuery;
```

Comment:
It took me a moment to realize this was not the same variable as localShardSqlQuery